MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c

```diff
@@ -0,0 +1,586 @@
+    NoDeallocAfterSanitize
+    ));
+
+  SanitizeCdw10Cdw11.Ndas   = NoDeallocAfterSanitize;
+  SanitizeCdw10Cdw11.Sanac  = SanitizeAction;
+  SanitizeCdw10Cdw11.Ovrpat = OverwritePattern;
+  CopyMem (&CommandPacket.NvmeCmd->Cdw10, &SanitizeCdw10Cdw11, sizeof (NVME_ADMIN_SANITIZE));
+
+  //
+  // Send Format NVM command via passthru and wait for completion
+  //
+  Status = Device->Controller->Passthru.PassThru (
+                                          &(Device->Controller->Passthru),
+                                          NamespaceId,
+                                          &CommandPacket,
+                                          NULL
+                                          );
+
+  if (EFI_ERROR (Status)) {
+    UINT16  StatusField = 0;
+    UINT16  Sct         = 0;
+    UINT16  Sc          = 0;
+
+    StatusField = (UINT16)((CommandPacket.NvmeCompletion->DW3 & NVME_CQE_STATUS_FIELD_MASK) >>
+                           NVME_CQE_STATUS_FIELD_OFFSET);
+
+    Sc  = (StatusField & NVME_CQE_STATUS_FIELD_SC_MASK) >> NVME_CQE_STATUS_FIELD_SC_OFFSET;
+    Sct = (StatusField & NVME_CQE_STATUS_FIELD_SCT_MASK) >> NVME_CQE_STATUS_FIELD_SCT_OFFSET;
+
+    DEBUG ((DEBUG_ERROR, "ERROR %a - %r\n", __FUNCTION__, Status));
+    DEBUG ((DEBUG_ERROR, "SCT = 0x%x, SC = 0x%x\n", Sct, Sc));
+
+    //
+    // Check for an error status code of "Invalid Command Opcode" in case
+    // the NVM Express controller does not support Sanitize. If the NVM
+    // Exress Controller does not support Sanitize, then send a Format NVM
+    // admin command instead to perform the Purge operation.
+    if ((Sct == NVME_CQE_SCT_GENERIC_CMD_STATUS) &&
+        (Sc == NVME_CQE_SC_INVALID_CMD_OPCODE))
+    {
+      UINT32  FnvmSes = 0;
+
+      switch (SanitizeCdw10Cdw11.Sanac) {
+        case SANITIZE_ACTION_BLOCK_ERASE:
+          FnvmSes = SES_USER_DATA_ERASE; // User Data Erase (LBAs indeterminate after)
+          break;
+        case SANITIZE_ACTION_CRYPTO_ERASE:
+          FnvmSes = SES_CRYPTO_ERASE; // Crypto Erase
+          break;
+        case SANITIZE_ACTION_OVERWRITE:
+        case SANITIZE_ACTION_EXIT_FAILURE_MODE:
+        default:
+          //
+          // Cannot perform an equivalent FormatNVM action/operation
+          //
+          FnvmSes = SES_NO_SECURE_ERASE;
+          break;
+      }
+
+      if ((FnvmSes == SES_USER_DATA_ERASE) || (FnvmSes == SES_CRYPTO_ERASE)) {
+        Status = NvmExpressFormatNvm (
+                   This,
+                   NVME_ALL_NAMESPACES,
+                   FnvmSes,
+                   Device->NamespaceData.Flbas
+                   );
+      }
+    }
+  }
+
+  return Status;
+}
+
+/**
+  Clear Media utilizes transport native WRITE commands to write a fixed pattern
+  of non-sensitive data to the media.
+
+  NOTE: The caller shall send buffer of one sector/LBA size with overwrite data.
+  NOTE: This operation is a blocking call.
+  NOTE: This function must be called from TPL aaplication or callback.

```

Comment:
Good question. Sean Brogan had similar feedback as well. When I asked him whether he wanted me to add a runtime check or just a comment in the code, he stated a comment was fine. Again, we can always add runtime checks at a later time if we feel the need.

Reply:
I think it makes sense to check it. There are places in the code that violate TPL calling restrictions but cannot be updated (easily) because so much other code now depends on it.